### PR TITLE
ENG-3598 Package command implementation

### PR DESCRIPTION
**Pull request by zonia3000**
In addition to package command implementation, I refactored some parts of the code:

* I split the `Stack` enum in `MicroFrontendStack` and `MicroServiceStack` and I defined a `StackFor<T extends ComponentType>` type that is used to avoid that a MFE stack can be assigned to a MS and vice-versa. This is needed for correctly defining the `CommandsDefinition` type, which is used by the command factory to dynamically obtain the commands to build/package a component.
* I changed the promise of `ProcessExecutorService` to resolve always, as happens in the parallel version, so we can avoid to catch the exception in the command in order to display a proper error message
* I created a `BaseBuildCommand` that can be used by build, build-all and package commands, since it contains some methods for handling execution of commands on a set of components. I know that there is a lot of logic in the command, but we have to display logs and progress bars while the bundle is packed.

### CODE REVIEW

File: src/commands/base-build.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+import * as path from 'node:path'
+import { Component, ComponentType } from '../models/component'
+import {
+  MICROFRONTENDS_FOLDER,
+  MICROSERVICES_FOLDER,
+  OUTPUT_FOLDER
+} from '../paths'
+import { mkdirSync } from 'node:fs'
+
+export abstract class BaseBuildCommand extends Command {
```
Comment (cecchisandrone) [inline src/commands/base-build.ts:23]: This is appearing in the command help, it should be hidden

File: src/commands/build.ts [hunk 2]
```diff
@@ -22,21 +22,18 @@ export default class Build extends Command {
...
-        } else {
-          console.error(
-            `Build failed, exit with code 1 and message ${error.message}`
-          )
-          this.exit(1)
-        }
+    const result = await componentService.build(args.name)
+    if (result !== 0) {
+      if (typeof result === 'number') {
+        console.error(`Build failed, exit with code ${result}`)
```
Comment (cecchisandrone) [inline src/commands/build.ts]: replace console.error() with this.error()

File: src/commands/base-build.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+    const executorService = new ParallelProcessExecutorService(executionOptions)
+
+    await this.parallelBuild(executorService, components)
+  }
+
+  private buildExecutionOptions(
+    components: Array<Component<ComponentType>>,
+    commandPhase: Phase
+  ) {
```
Comment (cecchisandrone) [inline src/commands/base-build.ts]: missing return type?

File: src/commands/base-build.ts [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+import {
+  MICROFRONTENDS_FOLDER,
+  MICROSERVICES_FOLDER,
+  OUTPUT_FOLDER
+} from '../paths'
+import { mkdirSync } from 'node:fs'
+
+export abstract class BaseBuildCommand extends Command {
+  public async buildAllComponents(commandPhase: Phase): Promise<void> {
+    this.log('Building packages...')
```
Comment (cecchisandrone) [inline src/commands/base-build.ts]: Building components...

File: src/commands/build.ts [hunk 5]
```diff
@@ -1,9 +1,9 @@
-import { CliUx, Command } from '@oclif/core'
+import { CliUx } from '@oclif/core'
 import { BundleService } from '../services/bundle-service'
 import { ComponentService } from '../services/component-service'
-import { ExecutionError } from '../services/process-executor-service'
+import { BaseBuildCommand } from './base-build'
 
-export default class Build extends Command {
+export default class Build extends BaseBuildCommand {
   static description = 'Build the component'
```
Comment (cecchisandrone) [inline src/commands/build.ts]: Build bundle components

File: src/commands/base-build.ts [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+      progress.update(progress.value + 1)
+    })
+
+    const results = await executorService.execute()
+    progress.stop()
+
+    this.checkBuildResults(results, components)
+  }
+
+  public checkBuildResults(
```
Comment (cecchisandrone) [inline src/commands/base-build.ts]: should this be `private`?

File: src/services/microfrontend-service.ts [hunk 7]
```diff
@@ -79,7 +86,7 @@ export class MicroFrontendService {
       throw new CLIError(`Directory ${mfedir} does not exist`)
     }
 
-    fs.rmSync(mfedir, { recursive: true, force: true });
+    fs.rmSync(mfedir, { recursive: true, force: true })
```
Comment (raxkaynan) [inline src/services/microfrontend-service.ts]: I think it's safe to remove `force: true` here.

File: src/commands/package.ts [hunk 8]
```diff
@@ -17,15 +28,47 @@ export default class Package extends Command {
...
+    BundleService.verifyBundleInitialized(bundleDir)
+
     const { flags } = await this.parse(Package)
-    await this.getDockerOrganization(flags.org)
+
+    const bundleDescriptorService = new BundleDescriptorService(bundleDir)
+    const bundleDescriptor = bundleDescriptorService.getBundleDescriptor()
+
+    const needsBuild = process.stdout.isTTY
+      ? await CliUx.ux.confirm('Rebuild components?')
```
Comment (cecchisandrone) [inline src/commands/package.ts]: Do you want to rebuild all bundle components?

File: src/commands/package.ts [hunk 9]
```diff
@@ -17,15 +28,47 @@ export default class Package extends Command {
...
+
+    if (needsBuild) {
+      await this.buildAllComponents(Phase.Package)
+    }
+
+    const dockerOrganization = await this.getDockerOrganization(flags.org)
+
+    await this.buildMicroServicesDockerImages(
+      dockerOrganization,
+      bundleDescriptor.version
```
Comment (cecchisandrone) [inline src/commands/package.ts]: Docker images for MS will get their component version, not the bundle version

**Review by raxkaynan — CHANGES_REQUESTED** [review]

File: src/commands/package.ts [hunk 10]
```diff
@@ -17,15 +28,47 @@ export default class Package extends Command {
     org: Flags.string({
       char: 'o',
       description: 'Docker organization name'
+    }),
+    file: Flags.string({
+      char: 'f',
+      description: 'Name of the Dockerfile (default is Dockerfile)',
```
Comment (cecchisandrone) [inline src/commands/package.ts]: Bundle Dockerfile (default is Dockerfile)

File: src/commands/package.ts (hunk 8 above)
Comment (cecchisandrone) [inline src/commands/package.ts]: Let's remove this as per the comment above

File: src/commands/package.ts [hunk 11]
```diff
@@ -54,4 +97,75 @@ export default class Package extends Command {
     )
     return newOrganization
   }
+
+  private async buildMicroServicesDockerImages(
```
Comment (cecchisandrone) [inline src/commands/package.ts]: I would check for Dockerfile before calling the docker command and fail silently with it. In this way we can give a more explicit error like "Dockerfile not found for microservice <abc>. Please provide one in order to proceed with the bundle packaging"

File: src/commands/package.ts [hunk 12]
```diff
@@ -54,4 +97,75 @@ export default class Package extends Command {
...
+    dockerOrganization: string,
+    tag: string
+  ) {
+    const componentService = new ComponentService()
+    const microServices = componentService.getComponents(
+      ComponentType.MICROSERVICE
+    )
+
+    this.log('Building microservices Docker images...')
```
Comment (cecchisandrone) [inline src/commands/package.ts]: Probably better to colorize macro steps or make it more visible (------ Building ... ----) in order to be more evident in the logs

File: src/commands/base-build.ts [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+    results: ProcessExecutionResult[],
+    components: Array<Component<ComponentType>>
+  ): void {
+    if (results.some(result => result !== 0)) {
+      let errorMessage = 'Following components failed to build:\n'
```
Comment (cecchisandrone) [inline src/commands/base-build.ts]: The following components failed to build:\n

File: src/commands/package.ts [hunk 14]
```diff
@@ -17,15 +28,47 @@ export default class Package extends Command {
     org: Flags.string({
       char: 'o',
       description: 'Docker organization name'
+    }),
+    file: Flags.string({
```
Comment (cecchisandrone) [inline src/commands/package.ts]: Let's modify a bit the UX. Add a flag -b to build all the components. If missing build will not be invoked

File: src/commands/package.ts [hunk 15]
```diff
@@ -54,4 +97,75 @@ export default class Package extends Command {
...
+
+    await this.parallelBuild(executorService, microServices)
+  }
+
+  private async buildBundleDockerImage(
+    bundleDir: string,
+    bundleDescriptor: BundleDescriptor,
+    dockerOrganization: string
+  ) {
+    CliUx.ux.action.start('Creating bundle package')
```
Comment (cecchisandrone) [inline src/commands/package.ts]: Let's also log the full bundle image name at the end

**Review by cecchisandrone — APPROVED** [review]

**Pull request merged**